paasta_tools/mesos_maintenance.py

```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    :param mesos_secrets: optional argument specifying the path to the file containing the mesos-slave credentials
+    :returns: a string containing the secret/password
+    """
+    return load_credentials(mesos_secrets)[1]
+
+
+def reserve(slave_id, resource, amount):
+    """Dynamically reserve resources in marathon to prevent tasks from using them.
+    :param slave_id: the id of the mesos slave
+    :param resource: the mesos resource type we want to reserve (i.e. cpus, mem, gpus, etc)
+    :param amount: integer representing the amount of the above resource to reserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically reserving %f %s on %s" % (amount, resource, slave_id))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(slave_id, resource, amount)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = reserve_api()
+    try:
+        reserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError as e:
+        e.msg = "Error adding dynamic reservation. Got error: %s" % e.msg
+        raise
+    return reserve_output
+
+
+def unreserve(slave_id, resource, amount):
+    """Dynamically unreserve resources in marathon to allow tasks to using them.
+    :param slave_id: the id of the mesos slave
+    :param resource: the mesos resource type we want to unreserve (i.e. cpus, mem, gpus, etc)
+    :param amount: integer representing the amount of the above resource to unreserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically unreserving %d %s on %s" % (amount, resource, slave_id))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(slave_id, resource, amount)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = unreserve_api()
+    try:
+        unreserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError as e:
+        e.msg = "Error adding dynamic unreservation. Got error: %s" % e.msg
+        raise
+    return unreserve_output
+
+
+def reserve_all_resources(hostnames):
+    mesos_state = get_mesos_state_summary_from_leader()
+    hosts = []
+    for hostname in hostnames:
+        if '|' in hostname:
+            (host, _) = hostname.split('|')
+            hosts.append(host)
+        else:
+            hosts.append(hostname)
+    for slave in mesos_state['slaves']:
+        if slave['hostname'] in hosts:
+            log.info("Reserving all resources on %s" % slave['hostname'])
+            slave_id = slave['id']
+
+            disk = slave['resources']['disk'] - slave['used_resources']['disk']
+            for role in slave['reserved_resources']:
+                disk -= slave['reserved_resources'][role]['disk']
+            reserve(slave_id=slave_id, resource='disk', amount=disk)
```

Comment:
@nhandler I don't quite get what you mean. we want to catch exceptions thrown by the ``reserve`` call here, so I was just expecting something like

```
try:
  reserve(slave_id=slave_id, resource='disk', amount=disk)
except HTTPError as e:
   log.debug('couldn't reserve resources; bailing')
   raise
```

(or whatever action you think is the *right* thing to do). [we rereaise the exception after it is caught in](https://github.com/Yelp/paasta/blob/master/paasta_tools/mesos_maintenance.py#L59) ``base_api`` and again in ``reserve`` which I think is fine; you need to catch the HTTPError thrown and deal with it appropriately at each call site.

> This will make it possible to adjust how we deal with exceptions for each of the API calls.

This isn't really useful - the answer is mostly going to be log.debug and raise - the handling of an exception will be unique at each call site, not at the api call